src/type-checker.cc

```diff
@@ -713,24 +719,34 @@ Result TypeChecker::OnLoop(const TypeVector& param_types,
   return result;
 }
 
-Result TypeChecker::OnMemoryCopy(const Limits& limits) {
-  return CheckOpcode3(Opcode::MemoryCopy, &limits, &limits, &limits);
+Result TypeChecker::OnMemoryCopy(const Limits& src_limits,
+                                 const Limits& dst_limits) {
+  std::optional<Type> src_type = GetLimitType(src_limits);
+  std::optional<Type> dst_type = GetLimitType(dst_limits);
+  std::optional<Type> size_type = Type(Type::I32);
+  // The memory64 proposal specifies that the type of the size argument should
+  // be the mimimum of the two memory types.
+  if (src_limits.is_64 && dst_limits.is_64) {
+    size_type = Type(Type::I64);
+  }
```

Comment:
It seems like we could perhaps avoid the rest of this change if we restructured this method a little bit.

How about something like this:

```
Limits size_limits = src_limits;
// The memory64 proposal specifies that the type of the size argument should
// be the mimimum of the two memory types.
if (src_limits.is_64 && !dst_limits.is_64) {
  size_limits = dst_limits;
}
return CheckOpcode3(Opcode::MemoryCopy, &src_limits, &dst_limits, &size_limits);
```

I think this would avoid change the majority of these function signatures and all the additional use of `std::optional`.

Do you think its makes more sense to be passing around `std::optional<Type>` everywhere instead of `Limits*`?  If so maybe we can seperate out that change and land that first as an NFC? (non-functional change)

Reply:
Yes, I can definitely split them up if you prefer. I was thinking `std::optional<Type>` made more sense then `Limits*` because ultimately the `Type` is the only field in `Limits` that was being used.

Reply:
I created https://github.com/WebAssembly/wabt/pull/2032 which does just the `Limits* -> std::optional<Type>` change. And I'll rebase this PR on top of it so that it just contains the change in logic.

Reply:
I already rebased https://github.com/WebAssembly/wabt/pull/2032 and this on top of https://github.com/WebAssembly/wabt/pull/2031.

Reply:
Just to be clear: This PR now has two commits, but the first commit is it's own PR as https://github.com/WebAssembly/wabt/pull/2032.